From 0103e7010746824610fe8c30fc1345fcc16bcdd9 Mon Sep 17 00:00:00 2001
From: Wang Hai <wanghai38@huawei.com>
Date: Mon, 22 Aug 2022 20:08:56 +0800
Subject: [PATCH 025/261] [Huawei] thermal/drivers/core: Revert "[Backport]
 thermal/drivers/core: Use a char pointer for the cooling device name

Offering: HULK
hulk inclusion
category: bugfix
bugzilla: 187506

--------------------------------

This reverts commit 5674d780b865b658196fd06d5ecd86aadc72b310.

commit 5674d780b865 ("[Backport] thermal/drivers/core: Use a char pointer
for the cooling device name") will break KABI and cannot be fixed. It
is not a bugfix, so revert it.

Fixes: 5674d780b865 ("[Backport] thermal/drivers/core: Use a char pointer for the cooling device name")
Signed-off-by: Wang Hai <wanghai38@huawei.com>
---
 .../ethernet/mellanox/mlxsw/core_thermal.c    |  2 +-
 drivers/thermal/thermal_core.c                | 40 ++++++++-----------
 include/linux/thermal.h                       |  2 +-
 3 files changed, 19 insertions(+), 25 deletions(-)

diff --git a/drivers/net/ethernet/mellanox/mlxsw/core_thermal.c b/drivers/net/ethernet/mellanox/mlxsw/core_thermal.c
index ecd1856bef5e..7ec1d0ee9bee 100644
--- a/drivers/net/ethernet/mellanox/mlxsw/core_thermal.c
+++ b/drivers/net/ethernet/mellanox/mlxsw/core_thermal.c
@@ -133,7 +133,7 @@ static int mlxsw_get_cooling_device_idx(struct mlxsw_thermal *thermal,
 	/* Allow mlxsw thermal zone binding to an external cooling device */
 	for (i = 0; i < ARRAY_SIZE(mlxsw_thermal_external_allowed_cdev); i++) {
 		if (strnstr(cdev->type, mlxsw_thermal_external_allowed_cdev[i],
-			    strlen(cdev->type)))
+			    sizeof(cdev->type)))
 			return 0;
 	}
 
diff --git a/drivers/thermal/thermal_core.c b/drivers/thermal/thermal_core.c
index 5e48e9cfa044..b906aada4876 100644
--- a/drivers/thermal/thermal_core.c
+++ b/drivers/thermal/thermal_core.c
@@ -1092,7 +1092,10 @@ __thermal_cooling_device_register(struct device_node *np,
 {
 	struct thermal_cooling_device *cdev;
 	struct thermal_zone_device *pos = NULL;
-	int ret;
+	int result;
+
+	if (type && strlen(type) >= THERMAL_NAME_LENGTH)
+		return ERR_PTR(-EINVAL);
 
 	if (!ops || !ops->get_max_state || !ops->get_cur_state ||
 	    !ops->set_cur_state)
@@ -1102,17 +1105,14 @@ __thermal_cooling_device_register(struct device_node *np,
 	if (!cdev)
 		return ERR_PTR(-ENOMEM);
 
-	ret = ida_simple_get(&thermal_cdev_ida, 0, 0, GFP_KERNEL);
-	if (ret < 0)
-		goto out_kfree_cdev;
-	cdev->id = ret;
-
-	cdev->type = kstrdup(type ? type : "", GFP_KERNEL);
-	if (!cdev->type) {
-		ret = -ENOMEM;
-		goto out_ida_remove;
+	result = ida_simple_get(&thermal_cdev_ida, 0, 0, GFP_KERNEL);
+	if (result < 0) {
+		kfree(cdev);
+		return ERR_PTR(result);
 	}
 
+	cdev->id = result;
+	strlcpy(cdev->type, type ? : "", sizeof(cdev->type));
 	mutex_init(&cdev->lock);
 	INIT_LIST_HEAD(&cdev->thermal_instances);
 	cdev->np = np;
@@ -1122,9 +1122,13 @@ __thermal_cooling_device_register(struct device_node *np,
 	cdev->devdata = devdata;
 	thermal_cooling_device_setup_sysfs(cdev);
 	dev_set_name(&cdev->device, "cooling_device%d", cdev->id);
-	ret = device_register(&cdev->device);
-	if (ret)
-		goto out_kfree_type;
+	result = device_register(&cdev->device);
+	if (result) {
+		thermal_cooling_device_destroy_sysfs(cdev);
+		ida_simple_remove(&thermal_cdev_ida, cdev->id);
+		put_device(&cdev->device);
+		return ERR_PTR(result);
+	}
 
 	/* Add 'this' new cdev to the global cdev list */
 	mutex_lock(&thermal_list_lock);
@@ -1142,15 +1146,6 @@ __thermal_cooling_device_register(struct device_node *np,
 	mutex_unlock(&thermal_list_lock);
 
 	return cdev;
-
-out_kfree_type:
-	thermal_cooling_device_destroy_sysfs(cdev);
-	kfree(cdev->type);
-	put_device(&cdev->device);
-out_ida_remove:
-	ida_simple_remove(&thermal_cdev_ida, cdev->id);
-out_kfree_cdev:
-	return ERR_PTR(ret);
 }
 
 /**
@@ -1309,7 +1304,6 @@ void thermal_cooling_device_unregister(struct thermal_cooling_device *cdev)
 	ida_simple_remove(&thermal_cdev_ida, cdev->id);
 	device_del(&cdev->device);
 	thermal_cooling_device_destroy_sysfs(cdev);
-	kfree(cdev->type);
 	put_device(&cdev->device);
 }
 EXPORT_SYMBOL_GPL(thermal_cooling_device_unregister);
diff --git a/include/linux/thermal.h b/include/linux/thermal.h
index 7097d4dcfdd0..176d9454e8f3 100644
--- a/include/linux/thermal.h
+++ b/include/linux/thermal.h
@@ -92,7 +92,7 @@ struct thermal_cooling_device_ops {
 
 struct thermal_cooling_device {
 	int id;
-	char *type;
+	char type[THERMAL_NAME_LENGTH];
 	struct device device;
 	struct device_node *np;
 	void *devdata;
-- 
2.17.1

